bestofjs/bestofjs
# Integrate e2e with playwright

## Goal

Add e2e using [playwright](https://playwright.dev/docs/intro) for local and CI
Added some e2e tests to sanity check some important parts of the home page, and test navigation using the header items

The CI workflow has changed as follows.
> Unit tests & Lint >> Build and Deploy to Vercel >> Test E2E on vercel deployment

Notes:
- Will run automatically for PR's on `develop` branch only.
- Only using chromium to test for speed and stability until we have a reason to test using a different browser
- The `More` header dropdown items are causing problems as clicking them using playwright will take you to  `/about `instead of opening the dropdown so this is not yet implemented.

## Quirks
When running tests sometimes you will see the nextjs web server throw errors like this
```
TypeError [ERR_INVALID_STATE]: Invalid state: ReadableStream is already closed
```

Although this does not affect our tests or functionality in the end,
this seems to be an issue when using ` Streaming Server Rendering` features such as `Suspense` or `loading.tsx`

https://github.com/vercel/next.js/issues/55608


## How to test

[See README `Testing` section here or follow instructions below](https://github.com/bestofjs/bestofjs/blob/4ac976d1c70e067b4e450b5391465529f4258801/apps/bestofjs-nextjs/README.md#testing)

1. Confirm `vitest` only runs unit tests
```
pnpm test
```

2. Install playwright for e2e testing
```
pnpm test:e2e:install
```

3. Test using next production server
 
Playwright will automatically start the production server so just make sure to create a next build

```bash
cd apps/bestofjs-nextjs/
pnpm build && pnpm test:e2e
```

## Screenshots

Testing with vitest
![image](https://github.com/bestofjs/bestofjs/assets/2955134/b4cde33b-79c7-4b15-a9ac-d1f259413144)

Testing with playwright
![image](https://github.com/bestofjs/bestofjs/assets/2955134/e0000e58-983f-4ce0-946d-c94fd92fa77b)

Updated CI workflow
![image](https://github.com/bestofjs/bestofjs/assets/2955134/c4a01b49-7ed4-4512-a7c2-ea2e6ae4c27d)
